### DISCUSSION TITLE
Maven on plugin RCP app -> Unable to locate feature 'org.eclipse.equinox.executable'. This feature is required for native product launchers

### DISCUSSION THREAD

**Discussion by jclausius-sg** [eclipse-tycho/tycho#4443] [Q&A] [answered]
I have an older Plugin based Eclipse project (non e4) that over the years has been brought along.  

The workspace was adapted to use a maven build a few years back where there is a main "app.build.root" project where it's pom.xml has been updated to use the Eclipse 2024-09 P2 repository and Tycho 4.0.9.  App.build.root's pom also references all of the various plugins that comprise the app in addition to a app.build.product project that has just regular app.product files as &lt;modules&gt;.

&lt;modules&gt;
    &lt;module&gt;../app.core&lt;/module&gt;
    &lt;module&gt;../app.jni&lt;/module&gt;
    &lt;module&gt;../app..util&lt;/module&gt;
    &lt;module&gt;../app.build.product&lt;/module&gt;
&lt;/modules&gt;

Note, there are also no *.target files in app.build.product, just the Product file generated by Eclipse.

( Also note, in the app.build.root's pom.xml flie, a plugin for "target-platform-configuration" is used with the &lt;configuration&gt; &lt;enviornments&gt; of  { linux, gtk, x86_64 }, { macosx, cocoa, x86_64 }, {macosx, cocoa, aarch64 }, {win32, win32, x86_64 }. We've never got this to work on any platform, but bringing it just in case. )

In regards to the app.product file, it lists all the plugins required for the app and also has checked the "The product includes native launcher artifacts."

In any case, after updating the pom.xml's to the latest libraries/versions, the maven build fails with the following:

[ERROR] Failed to execute goal org.eclipse.tycho:tycho-p2-publisher-plugin:4.0.9:publish-products (default-publish-products) on project app.build.product: Publisher failed. Verify your target-platform-configuration and executionEnvironment are suitable for proper resolution: Unable to locate feature 'org.eclipse.equinox.executable'. This feature is required for native product launchers. -&gt; [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn &lt;args&gt; -rf :app.build.product
----

Now org.eclipse.equinox.executable is a feature in the Eclipse P2 repository. I tried adding a &lt;dependency-resolution&gt; to the target-platform-configuration plugin's &lt;configuration&gt;, but that seemed to create other problems like :
"Missing requirement: app.product 1.0.0.qualifier requires 'org.eclipse.equinox.p2.iu; org.eclipse.equinox.p2.iu 0.0.0' but it could not be found"

----
Any ideas or information I can provide to get to the bottom of this would be greatly appreciated.
-Jeff

**Comment by jclausius-sg** [comment]
A quick addendum - I know this probably has nothing to do with a tycho build because it is using the Eclipse IDE, but if I open the app.product file and click Launch as Eclipse Executable or Export it, I can generated an executable binary, so it is not believed the problem is in that file.

-Jeff

**Comment by laeubi** [comment]
This is usually a configuration problem so it is quite hard to tell with just some fragments of code, as you are seeing other dependency revolution problems it looks like your target configuration is somehow messed up and the build does not see the required repositories.

**Reply by jclausius-sg** [reply]
Thank you.  Some progress was made here.

First off, I had thought I did this, but apparently did not commit changes to version control.  In any case, I:
- Opened Eclipse with the project.  
- Opened the app.product file with the Product Configuration Editor
- Went to the Contents tab and Removed ALL of the plugins (the product configuration is based on plugins).
- Using the "Add" button, re-added all my plugins for this specific app
- Next, clicked Add Required Plug-ins so Eclipse would add the rest of the plugins required for RCP and used by my App plugins.
- Saved the file
- Next went to the Configuration tab, clicked "Add recommended" to the Start Levels
- Saved the file

Apparently org.eclipse.equinox.executable and org.eclipse.equinox.p2.iu were considered plugins for past versions of Eclipse, as with the process above they are now removed from the app.product's plugin section.

---
A run of maven now produces:

[ERROR] Failed to execute goal org.eclipse.tycho:tycho-p2-publisher-plugin:4.0.9:publish-products (default-publish-products) on project app.build.product: Publisher failed. Verify your target-platform-configuration and executionEnvironment are suitable for proper resolution: Cannot resolve dependencies of product app. See log for details. -&gt; [Help 1]
---
Within my main (in the build.root project), there is no &lt;executionEnvironment&gt; section defined for the &lt;target-platform-configuration&gt; section.  The tycho help just states, "Force an execution environment for dependency resolution."

So not certain what this particular error means or if I need to add a &lt;executionEnvironment&gt; section to the pom.xml of the build.root project.

Any additional guidance on this one would be greatly appreciated.

-Jeff

**Reply by jclausius-sg** [reply]
Oops this is probably important...

I looked up higher in the output.  The maven output also has the following:

[ERROR]   eclipse-plugin artifact with ID "org.eclipse.core.filesystem.win32.x86_64" and version matching "0.0.0" was not found in the target platform
[ERROR]   eclipse-plugin artifact with ID "org.eclipse.core.resources.win32.x86_64" and version matching "0.0.0" was not found in the target platform
[ERROR]   eclipse-plugin artifact with ID "org.eclipse.equinox.security.win32.x86_64" and version matching "0.0.0" was not found in the target platform
-----
I'm building on Linux and trying to target both Linux and Windows.  The app.build.root's pom.xml has the following:

&lt;plugin&gt;
&lt;groupId&gt;org.eclipse.tycho&lt;/groupId&gt;
&lt;artifactId&gt;target-platform-configuration&lt;/artifactId&gt;
&lt;version&gt;${tycho-version}&lt;/version&gt;
&lt;configuration&gt;
  &lt;environments&gt;
    &lt;environment&gt;
      &lt;os&gt;linux&lt;/os&gt;
      &lt;ws&gt;gtk&lt;/ws&gt;
      &lt;arch&gt;x86_64&lt;/arch&gt;
    &lt;/environment&gt;
    &lt;environment&gt;
      &lt;os&gt;win32&lt;/os&gt;
      &lt;ws&gt;win32&lt;/ws&gt;
      &lt;arch&gt;x86_64&lt;/arch&gt;
    &lt;/environment&gt;
  &lt;/environments&gt;
&lt;/configuration&gt;
&lt;/plugin&gt;

---
Is there something else that is needed to grab binaries for other operating systems?

-Jeff

**Comment by mickaelistria** [comment]
Is it the first experience you're having with Tycho? Migrating an old code base can be a tricky exercise, before starting it, make sure you've gone though a few tutorials to understand the underlying concepts and error messages; this is pre-requisite knowledge for being successful here.

**Comment by jclausius-sg** [comment, marked as answer]
A follow up to this problem.  

The pom.xml and app.product file were last build against Eclipse 2021-12, but were updated to Eclipse 2024-09.

While I had updated the Eclipse p2 repository in the pom.xml file, and had let Eclipse 2024-09 make changes to app.product, I had not committed the changes to app.product's listing of required &lt;plugins&gt; to our version control system.  Once that change was committed, the build for maven started working again.

Thanks for everyone's suggestions. It was laeubi's suggestion that pointed me in the right direction.

Thank you,
-Jeff
